funsor/contract.py

```diff
@@ -0,0 +1,242 @@
+
+
+@eager.register(Contract, Tensor, Tensor, frozenset)
+def eager_contract_tensor_tensor(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    if reduced_vars:
+        out_vars = (frozenset(lhs.inputs) | frozenset(rhs.inputs)) - reduced_vars
+    else:
+        out_vars = {}
+
+    out_inputs = OrderedDict([
+        (v, lhs.inputs[v] if v in lhs.inputs else rhs.inputs[v]) for v in out_vars])
+
+    return Tensor(
+        opt_einsum.contract(lhs.data, list(lhs.inputs.keys()),
+                            rhs.data, list(rhs.inputs.keys()),
+                            list(out_vars), backend="torch"),
+        out_inputs
+    )
+
+
+@optimize.register(Contract, Funsor, Reduce, frozenset)
+def contract_reduce(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    # XXX should we be doing this conversion at all given that Reduce
+    # is already handled by the optimizer?
+    if rhs.op is ops.add:
+        base_lhs = _make_base_lhs(rhs.arg, rhs.reduced_vars, normalized=False)
+        inner = Contract(base_lhs, rhs.arg, rhs.reduced_vars)
+        return Contract(lhs, inner, reduced_vars)
+    return None
+
+
+@optimize.register(Contract, ATOMS, Finitary, frozenset)
+def contract_ground_finitary(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    # exploit linearity of integration
+    if rhs.op is ops.add:
+        return Finitary(
+            ops.add,
+            tuple(Contract(lhs, operand, reduced_vars) for operand in rhs.operands)
+        )
+
+    # pull out constant terms that do not depend on the lhs
+    if rhs.op is ops.mul:
+        constants, new_operands = _find_constants(lhs, rhs.operands, reduced_vars)
+        if new_operands and constants:
+            # this term should equal Finitary(mul, constants) for probability lhss
+            outer = Finitary(ops.mul, constants)
+            inner = Contract(lhs, Finitary(rhs.op, new_operands), reduced_vars)
+            return outer * inner
+        elif not new_operands and constants:
+            return Finitary(ops.mul, constants)
+
+    return None
+
+
+@optimize.register(Contract, Finitary, (Finitary,) + ATOMS, frozenset)
+def contract_finitary_ground(lhs, rhs, reduced_vars):
+    result = _simplify_contract(lhs, rhs, reduced_vars)
+    if result is not None:
+        return result
+
+    # recursively apply law of iterated expectation
+    assert len(lhs.operands) > 1, "Finitary with one operand should have been passed through"
+    if lhs.op is ops.mul:
+        root_lhs, remaining_lhs = _order_lhss(lhs, reduced_vars)
+        if remaining_lhs is not None:
+            inner = Contract(remaining_lhs, rhs,
+                             reduced_vars & frozenset(remaining_lhs.inputs))
```

Comment:
Hmm, should this be
```py
reduced_vars.intersection(remaining.lhs.inputs).difference(root_lhs.inputs)
```
I would have guessed that a test like `"a,a,a->"` would have caught this 😕

Reply:
`_order_lhss` guarantees that this intersection is empty